### PR TITLE
HIP back end to GPU package to support AMD GPUs via the ROCm toolkit

### PR DESCRIPTION
**Pull request by Vsevak**
**Summary**
This PR is an update for the changes proposed in #1590. It has been merged with the current 'master' and now includes the changes from #1752, #1776, etc. Also, the building instructions in lib/gpu/README has been extended accordingly.

The PR enhances LAMMPS with the capability for using the new HIP backend for GPU-offloading of MD calculations as implemented in the GPU package. The GPU package has been developed originally with the support of CUDA and OpenCL backends using the  geryon library as the unification layer. This PR adds AMD HIP as a third backend to the geryon library. 
The AMD ROCm open source GPU framework provides the driver and the programming tools for new high performance AMD GPUs (including the new Vega20 architecture). ROCm supports OpenCL 2.0 standard and OpenCL 1.2 compatibility mode. However the existing OpenCL back-end of LAMMPS is not compatible with ROCm due to some deviations from the standard (#1368). Some small changes in all *.cu files are required. That is why the current version of GPU package can not be used with ROCm at all. This PR removes this restriction. Besides, the HIP backend is more than 30% faster than the OpenCL backend.
HIP is a part of the AMD ROCm framework. This new HIP-backend can be used either with the hcc compiler for AMD GPUs, or with the nvcc compiler for Nvidia GPUs. In the latter case, the HIP acts as a thin layer between the library and CUDA.

**Related Issues**
This PR is related to issue #1368 and opens a new way of running LAMMPS on AMD GPUs that support ROCm without fixing the current version of the GPU package OpenCL-backend that is not fully compatible with the OpenCL standard required for ROCm. The issue #1368 could be probably closed after merging.

**Author(s)**
Evgeny Kuznetsov (@eokuznetsov), Vladimir Stegailov (v.stegailov@hse.ru, stegailov@gmail.com, @vvsteg), Vsevolod Nikolskiy (thevsevak@gmail.com, @Vsevak)
International Laboratory for Supercomputer Atomistic Modelling and Multi-scale Analysis, National Research University Higher School of Economics, Russia

**Licensing**
By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**
This PR should not break backward compatibility. The solution is implemented as the third backend in addition to existing CUDA and OpenCL. The decision between CUDA, OpenCL, and HIP is a compile-time decision for compiling the library in lib/gpu.

**Implementation Notes**
See #1590.
This PR has been tested using ROCm 2.2-3.0 and different standard LAMMPS input examples on Radeon VII (gfx906), Radeon RX 480 (gfx803), Radeon RX Vega (gfx900),  and Titan V (sm_70) with OpenMPI and MPICH.
~~The new HIP backend has not been integrated into CMake yet.~~ (update by @rbberger)

**Post Submission Checklist**
- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [x] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

### CODE REVIEW

**Comment by akohlmey** [comment]
Thanks for the submission. We will look into it after the stable release and when we have access to suitable software/hardware for testing.

File: lib/gpu/Makefile.hip [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+	$(HIP_HOST_CC_CMD) -o $@ -c $< -I$(OBJ_DIR) $(HIP_GPU_SORT_ARCH) 
+
+$(OBJ_DIR)/lal_%.o: lal_%.cpp $(CUHS) $(ALL_H)
+	$(HIP_HOST_CC_CMD) -o $@ -c $< -I$(OBJ_DIR)
+
+# libgpu building
+
+$(LIB_DIR)/libgpu.a: $(OBJS)
+	$(AR) -crs $@ $(OBJS)
+	echo "export HIP_PLATFORM := $(HIP_PLATFORM)\n$(HIP_LIBS_TARGET)" > 'Makefile.lammps'
```
Comment (rbberger) [inline lib/gpu/Makefile.hip]: `echo` doesn't work here. It produces a single line with `\n` not being expanded to a newline. Use `printf` instead.

Comment (Vsevak) [reply lib/gpu/Makefile.hip]: Thank you, fixed.

**Comment by akohlmey** [comment]
Now that the stable release is out, this PR is under consideration to be merged. However, it looks we still need support for it in the CMake scripts and also it needs added documentation in the manual.

File: lib/gpu/Makefile.hip [hunk 2]
```diff
@@ -137,7 +137,7 @@ $(OBJ_DIR)/lal_%.o: lal_%.cpp $(CUHS) $(ALL_H)
 
 $(LIB_DIR)/libgpu.a: $(OBJS)
 	$(AR) -crs $@ $(OBJS)
-	echo "export HIP_PLATFORM := $(HIP_PLATFORM)\n$(HIP_LIBS_TARGET)" > 'Makefile.lammps'   
+	echo -e "export HIP_PLATFORM := $(HIP_PLATFORM)\n$(HIP_LIBS_TARGET)" > 'Makefile.lammps'
```
Comment (arghdos) [inline lib/gpu/Makefile.hip]: I recommend this be changed to printf, as `echo -e` is [not portable](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/echo.html#tag_20_37_16).  I noticed this as compiling fails on Ubuntu 16.04

**Comment by rbberger** [comment]
Added some basic CMake support. Still need to look into proper way of doing options.
It currently compiles, but I haven't done any functional tests.

Tested on the ROCm container (https://github.com/lammps/lammps/blob/master/tools/singularity/ubuntu18.04_amd_rocm.def)

```bash
mkdir build
cd build

export LC_ALL=C
export HCC_AMDGPU_TARGET=gfx906
cmake -D CMAKE_CXX_COMPILER=hipcc -D PKG_GPU=on -D GPU_API=HIP ../cmake
make -j 12
```

**Comment by rbberger** [comment]
Pretty much reached a dead end here. I don't see how I can make CMake work with `HIP_PLATFORM=nvcc` and `CMAKE_CXX_COMPILER=hipcc`. I've committed what I have so far.

**Comment by vvsteg** [comment]
> Pretty much reached a dead end here. I don't see how I can make CMake work with `HIP_PLATFORM=nvcc` and `CMAKE_CXX_COMPILER=hipcc`. 

Certainly, it would be very nice to have a possibility of using the HIP back-end with `nvcc`. But this capability looks redundant since there is still the original CUDA back-end.

As far as I understand, having both CUDA and HIP back-ends does not mean any overhead for new developments, since in the GPU package the algorithms are expressed using the `geryon` library unification layer.

**Comment by arghdos** [comment]
So, the way I've worked around using a different compiler for host-only API files in the past is to run find_package(HIP) to find the path to the hipconfig script (it's essentially $HIP_ROPT_DIR/../bin/hipconfig if I recall correctly). From there you can obtain the flags you need for compiling host only API files by executing hipconfig -C and storing it in your CMake flags.  Then set all the .cpp files to use the fake host target flags, while you can use the real hip targets for .cu files and the linking step (or not, if you guys are doing separate device linking). I'll note that this method works equally well for the HIP+AMD backend as well.

It's a real weak point in our CMake right now, and I'm going to see if we can add native support for a host-only API target in the future.

**Comment by rbberger** [comment]
I've managed to compile everything using the regular host compiler and hipcc for the cubins. So I'm not setting CMAKE_CXX_COMPILER in that case. For this to work, I needed to manually set the include and link paths with what I got out of `hipconfig -C`. Next up, testing if these binaries are actually usable and writing docs.

```bash
mkdir build
cd build

export CUDA_PATH=/usr/local/cuda
export HIP_PLATFORM=nvcc
cmake -D PKG_GPU=on -D GPU_API=HIP ../cmake
make -j 12
```

**Pull request merged**